packages/std/src/types.rs

```diff
@@ -109,3 +109,18 @@ pub struct MessageInfo {
 pub struct ContractInfo {
     pub address: Addr,
 }
+
+/// The structure contains additional information related to the
+/// contract's state migration procedure - the sender address and
+/// the contract's state version currently stored on the blockchain.
+/// The `old_state_version` is optional, since there is no guarantee
+/// that the currently stored contract's binary contains that information.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct MigrateInfo {
+    /// Address of the sender.
+    pub sender: Addr,
+    /// Version of the previous contract's state. It's optional, since
+    /// adding the state's version number to the binary is not a
+    /// mandatory feature.
+    pub old_state_version: Option<u64>,
+}
```

Comment:
Since the new version is hardcoded in the contract binary, the contract inherently "knows" it, but it also wouldn't hurt to provide it here again, so you have easy access to it.

Reply:
How can the contract currently access its own migrate version? Maybe it is good enough if there is a nice and simple way to do that

Reply:
You could access the `__CW_MIGRATE_VERSION` generated by the macro directly, but that's only available on the `wasm32` target and a bit of a hack relying on implementation detail.

Reply:
The new version is always provided by the developers in the code in the macro call for e.g.:
```rust
#[entry_point]
#[migrate_version(42)]
pub fn migrate(
    deps: DepsMut,
    _env: Env,
    msg: MigrateMsg,
    _migrate_info: MigrateInfo,
) -> Result<Response, HackError> {
    // [ ... ]
}
```

So even if we add this new version to `MigrateInfo` then it'd be always the same number that was picked in `migrate_version` macro. Is there any other scenario possible here?

Reply:
I was thinking about the need to mock the number in tests. But I think you always want to test the current migrate version against something old. So I don't see the need to have thick mockable.

I am not a huge fan of adding more globals as this easily harms composability of contracts. By the way, is this already a problem when adding a contracts with migrate version to another contract with migrate version? Doesn't that give you two globals? How would the above code snippet look when used with the library feature?

Reply:
It's a static variable with a `#[link_section = "cw_migrate_version"]` attribute. I don't know how it would be handled if you had multiple of them in the source tree, but the attribute is only defined if you use the `entry_point` macro, so the compiler will force you to guard both of them or none:
```rust
#[cfg_attr(not(feature = "library"), entry_point)]
#[cfg_attr(not(feature = "library"), migrate_version(42))]
pub fn migrate(deps: DepsMut, _env: Env, msg: MigrateMsg) -> Result<Response, HackError> {
```

Reply:
okay, so we can say `#[cfg_attr(not(feature = "library"), migrate_version(42))]` generates not only

```rust
static __CW_MIGRATE_VERSION: [u8; 2] = *"42"
```

but instead

```
#[allow(unused)]
#[doc(hidden)]
#[cfg(target_arch = "wasm32")]
#[link_section = "cw_migrate_version"]
static __CW_MIGRATE_VERSION: [u8; 2] = *"42"

// also available in unit tests
fn migrate_version() -> u64 {
    42u64
}
```

which you can then call as a free function in the `migrate` implementation?

Reply:
Yes, that is something we can do, but do you have a concrete case in mind where that is needed?

Reply:
Oh, now I get the point. My bad, somehow I completely forgot that we allow downgrades here. Then I agree we should have a way to get the version. The `fn migrate_version()` above looks fine to me.

Reply:
The new const method looks good - I'll add that to this PR

Reply:
Done

Reply:
Update: There is a problem with this approach. Let's say that user will add:
```rust
#[cfg_attr(not(feature = "library"), entry_point)]
#[cfg_attr(not(feature = "library"), migrate_version(42))]
pub fn migrate(deps: DepsMut, _env: Env, msg: MigrateMsg) -> Result<Response, HackError> {
    migrate_version()
    // [...]
```

This can cause a compilation error, when `library` is set - `migrate` method expects `migrate_version` which is not generated

Reply:
Oh, I really like that idea! Then the user can control it themselves and it doesn't feel so much like magic with a function appearing out of nowhere.

Reply:
Yes it would be a nice feature, but I couldn't find any solution for a compile time parsing `u64` into `[u8; X]`. Other solution would be to parse the constant as well, but imo it'd be terrible for users to have another place to add a new macro call. Third way I thought of was something like this:
```rust
#[cosmwasm_std::migrate_version]
const CONTRACT_VERSION: u64 = 33;
```

So instead of an attribute to the migrate entry point, we could add a new proc-macro called `migrate_version` that would parse such a user-defined const.

Reply:
True, I don't see a way to do that either. That's a bit unfortunate.

I would prefer to avoid adding another way to do the same thing. We already released a version with the current attribute to migrate, so we need to continue to support that.
Also not having it on the entrypoint opens up the can of worms of what happens when you specify multiple `migrate_version`s.
> Other solution would be to parse the constant as well

This made me think of allowing a second variant of `migrate_version` like this:
```rust
#[migrate_version(CONTRACT_VERSION = 42)]
fn migrate( ... )
```
That would then generate a constant with that name, as well as the `link_section` stuff. And I think it would feel more organic to then use `CONTRACT_VERSION` after that.

Reply:
Ok, now `#[migrate_version(CONTRACT_VERSION = 42)]` is also supported by entrypoints

Reply:
Still there is a problem with this approach - if the `library` feature flag is set, the constant is not generated. User gets compile error if the constant is used in `migrate` function

Reply:
We'd need to make the `migrate_version` a new proc-macro instead of an attribute for `entrypoint` one.

Reply:
Hmm, I see. We need to handle the following cases:
* entrypoint + migrate_version: constant and linker section generated
* migrate_version only: only the constant generated (to avoid conflicts with migrate version of contract that imports this one as library)

Can we check if the `entrypoint` macro is also specified inside the `migrate_version` macro?

Reply:
It's possible to check for migrate_version in entrypoint and vice versa, but I think it would be more tricky in the macro code and there would be problems with macros order. I **really** wanted to avoid that, but I prepared my own const-time function to cast `u64` into `[u8; N]`. I'll check rustc version compatibility, as we probably don't want users to enforce using only the newest compiler versions. I will push changes in a while

Reply:
With that new method user could write something like this:
```rust
const VERSION: u64 = 66;

#[cfg_attr(feature = "library", entry_point)]
#[cfg_attr(feature = "library", migrate_version(VERSION))]
pub fn migrate(
    deps: DepsMut,
    _env: Env,
    msg: MigrateMsg,
    _migrate_info: MigrateInfo,
) -> Result<Response, HackError> {
    if let Some(old_version) = _migrate_info.old_migrate_version {
        if VERSION < old_version {
            // [ ... ]
        }
    }
    // [ ... ]
}
```

Reply:
I was also thinking about that before, but as far as I can see, what is embedded in the linker section is the string representation, not just the underlying bytes of the `u64`, so `66` becomes `[54, 54]`.
I didn't think you can even have a const fn for that.

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c586d6f3a8c2968255366bc005364e4d

Reply:
Ok now it's done